File: leetcode/1510_Stone_Game_IV.py
from unittest import TestCase
from math import sqrt

class Solution:
    def winnerSquareGame(self, n: int) -> bool:
        dp = [False] * (n + 1)
        squares = [i * i for i in range(1, int(sqrt(n)) + 1)]
        for i in range(1, n + 1):
            for square in  squares:
                if square > i:
                    break
                elif not dp[i - square]:
                    dp[i] = True
                    break
        return dp[n]

class WinnerSquareGame(TestCase):
    def test_1(self):
        self.assertTrue(Solution().winnerSquareGame(1))

    def test_2(self):
        self.assertFalse(Solution().winnerSquareGame(2))

    def test_3(self):
        self.assertTrue(Solution().winnerSquareGame(4))                

File: leetcode/1510_Stone_Game_IV.rs
struct Solution {}

impl Solution {
    pub fn winner_square_game(n: i32) -> bool {
        let squares: Vec<u32> = (1..(n as f64).sqrt() as u32 + 1).map(|x| x * x).collect();
        let mut dp = vec![false; n as usize + 1];
        for i in 1..(n as u32 + 1) {
            for square in &squares {
                if *square > i {
                    break
                } else if ! dp[(i - *square) as usize] {
                    dp[i as usize] = true;
                    break
                }
            }
        }
        dp[n as usize]
    }
}


mod winner_square_game_test {
    use super::*;

    #[test]
    fn test_1() {
        assert!(Solution::winner_square_game(1));
    }

    #[test]
    fn test_2() {
        assert!(!Solution::winner_square_game(2));
    }
    
    #[test]
    fn test_3() {
        assert!(Solution::winner_square_game(4));
    }
}